IdentityCore/src/webview/background/ios/MSIDBackgroundTaskManager.m

```diff
@@ -68,52 +94,67 @@ + (MSIDBackgroundTaskManager *)sharedInstance
  */
 
-- (void)startOperationWithType:(MSIDBackgroundTaskType)type
+- (void)startOperationWithType:(MSIDBackgroundTaskType)type caller:(nonnull id)caller
 {
-    UIBackgroundTaskIdentifier backgroundTask = [self backgroundTaskWithType:type];
+    MSIDBackgroundTaskData *backgroundTaskData = [self backgroundTaskWithType:type];
     
-    if (backgroundTask != UIBackgroundTaskInvalid)
+    if (backgroundTaskData.backgroundTaskId != UIBackgroundTaskInvalid)
     {
         // Background task already started
         return;
     }
     
     MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Start background app task with type %ld", (long)type);
     
-    backgroundTask = [[MSIDAppExtensionUtil sharedApplication] beginBackgroundTaskWithName:@"Interactive login"
+    UIBackgroundTaskIdentifier backgroundTaskId = [[MSIDAppExtensionUtil sharedApplication] beginBackgroundTaskWithName:@"Interactive login"
                                                                   expirationHandler:^{
                                                                       MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Background task expired for type %ld", (long)type);
-                                                                      [self stopOperationWithType:type];
+                                                                      [self stopAllOperationsForType:type];
                                                                   }];
     
-    [self setBackgroundTask:backgroundTask forType:type];
+    [self setBackgroundTask:[[MSIDBackgroundTaskData alloc] initWithTaskId:backgroundTaskId caller:caller] forType:type];
 }
 
-- (void)stopOperationWithType:(MSIDBackgroundTaskType)type
+- (void)stopAllOperationsForType:(MSIDBackgroundTaskType)type
 {
-    UIBackgroundTaskIdentifier backgroundTask = [self backgroundTaskWithType:type];
+    MSIDBackgroundTaskData *backgroundTaskData = [self backgroundTaskWithType:type];
     
-    if (backgroundTask == UIBackgroundTaskInvalid)
+    if (backgroundTaskData.backgroundTaskId == UIBackgroundTaskInvalid)
     {
         // Background task not started
         return;
     }
     
     MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Stop background task with type %ld", (long)type);
-    [[MSIDAppExtensionUtil sharedApplication] endBackgroundTask:backgroundTask];
-    [self setBackgroundTask:UIBackgroundTaskInvalid forType:type];
+    [[MSIDAppExtensionUtil sharedApplication] endBackgroundTask:backgroundTaskData.backgroundTaskId];
+    [self setBackgroundTask:[[MSIDBackgroundTaskData alloc] initWithTaskId:UIBackgroundTaskInvalid caller:backgroundTaskData.callerReference] forType:type];
+}
+
+- (void)stopOperationWithType:(MSIDBackgroundTaskType)type caller:(nonnull id)caller
+{
+    MSIDBackgroundTaskData *backgroundTaskData = [self backgroundTaskWithType:type];
+    
+    if ((backgroundTaskData.backgroundTaskId == UIBackgroundTaskInvalid) || (![caller isEqual:backgroundTaskData.callerReference]))
```

Comment:
What does the ```isEqual``` refers to? Memory address?

Reply:
You're right, I had a different implementation in mind earlier and missed changing this back. isEqual will compare hash value instead of reference.

Reply:
Removed need to pass reference